protobuf/lib/src/protobuf/proto3_json.dart

```diff
@@ -81,32 +84,105 @@ Object? _writeToProto3Json(_FieldSet fs, TypeRegistry typeRegistry) {
+    // below
+    var value = fs._getField(fieldInfo.tagNumber);
+
+    bool skipField = true;
+    Object? jsonValue;
+    if (fieldInfo.isRepeated) {
+      if (context.emitDefaults || (value as List).isNotEmpty) {
+        skipField = false;
+        jsonValue = (value as PbListBase)
+            .map((element) => valueToProto3Json(element, fieldInfo.type))
+            .toList();
+      }
+    } else if (fieldInfo.isMapField) {
+      if (context.emitDefaults || (value as Map).isNotEmpty) {
+        skipField = false;
+        jsonValue = (value as PbMap).map((key, entryValue) {
+          var mapEntryInfo = fieldInfo as MapFieldInfo;
+          return MapEntry(convertToMapKey(key, mapEntryInfo.keyFieldType!),
+              valueToProto3Json(entryValue, mapEntryInfo.valueFieldType));
+        });
+      }
+    } else if (_isBytes(fieldInfo.type)) {
+      if (context.emitDefaults || (value as List).isNotEmpty) {
+        skipField = false;
+        if ((value as List).isEmpty) {
+          jsonValue = null;
+        } else {
+          jsonValue = valueToProto3Json(value, fieldInfo.type);
+        }
+      }
+    } else if (fieldInfo.isEnum) {
+      // For enums, the default value is the first value listed in the enum's type definition
+      final defaultEnum = fieldInfo.enumValues!.first;
+      if (context.emitDefaults ||
+          (value as ProtobufEnum).name != defaultEnum.name) {
+        skipField = false;
+        jsonValue = valueToProto3Json(value, fieldInfo.type);
+      }
+    } else if (fieldInfo.isGroupOrMessage) {
+      final originalValue = fs._values[fieldInfo.index!];
+      if (context.emitDefaults || originalValue != null) {
+        skipField = false;
+        if (originalValue == null) {
+          jsonValue = null;
+        } else {
+          jsonValue = valueToProto3Json(value, fieldInfo.type);
+        }
+      }
+    } else if (PbFieldType._baseType(fieldInfo.type) ==
+        PbFieldType._STRING_BIT) {
+      if (context.emitDefaults || (value as String).isNotEmpty) {
+        skipField = false;
+        jsonValue = valueToProto3Json(value, fieldInfo.type);
+      }
+    } else if (PbFieldType._baseType(fieldInfo.type) == PbFieldType._BOOL_BIT) {
+      if (context.emitDefaults || value != false) {
+        skipField = false;
+        jsonValue = valueToProto3Json(value, fieldInfo.type);
+      }
+    } else if (value is Int64) {
+      if (context.emitDefaults || !value.isZero) {
+        skipField = false;
+        jsonValue = valueToProto3Json(value, fieldInfo.type);
+      }
+    } else if (value is int) {
+      if (context.emitDefaults || value != 0) {
+        skipField = false;
+        jsonValue = valueToProto3Json(value, fieldInfo.type);
+      }
+    } else if (value is double) {
+      if (context.emitDefaults || value != 0.0) {
+        skipField = false;
+        jsonValue = valueToProto3Json(value, fieldInfo.type);
+      }
```

Comment:
Yes, this part was a bit wonky.  The reason I switched from value types to field types was because there are just so many numeric types to handle, so it became more readable to check the value types.  I think if we follow you top-level comment suggestion, then this is no longer an issue.  I'll wait for your and @sigurdm to reply on the best path forward there (if we go with the switch, from your other PR).